Kentico.Kontent.Delivery.Extensions.Autofac.DependencyInjection/DeliveryClientFactory.cs

```diff
@@ -0,0 +1,88 @@
+﻿using Autofac;
+using Autofac.Core.Registration;
+using Kentico.Kontent.Delivery.Abstractions;
+using Kentico.Kontent.Delivery.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json;
+using System;
+using System.Collections.Concurrent;
+
+namespace Kentico.Kontent.Delivery.Extensions.Autofac.DependencyInjection
+{
+    internal class DeliveryClientFactory : IDeliveryClientFactory
+    {
+        private readonly IOptionsMonitor<DeliveryOptions> _deliveryOptions;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly IComponentContext _componentContext;
+        private readonly ConcurrentDictionary<string, IDeliveryClient> _cache = new ConcurrentDictionary<string, IDeliveryClient>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DeliveryClientFactory"/> class.
+        /// </summary>
+        /// <param name="deliveryOptions">Used for notifications when <see cref="DeliveryOptions"/> instances change.</param>
+        /// <param name="serviceProvider">An <see cref="IServiceProvider"/> instance.</param>
+        /// <param name="componentContext">An autofac component context.</param>
+        public DeliveryClientFactory(IOptionsMonitor<DeliveryOptions> deliveryOptions, IServiceProvider serviceProvider,
+            IComponentContext componentContext)
+        {
+            _deliveryOptions = deliveryOptions;
+            _serviceProvider = serviceProvider;
+            _componentContext = componentContext;
+        }
+
+        /// <inheritdoc />
+        public IDeliveryClient Get(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (!_cache.TryGetValue(name, out var client))
+            {
+                var deliveryClientOptions = _deliveryOptions.Get(name);
+
+                // Validate that the option object is indeed configured
+                if (deliveryClientOptions.ProjectId != null)
```

Comment:
This check should be also obsolete with the transition to Autofac

Reply:
Removed

Reply:
The Autofac has no effects on how `_deliveryOptions `works

Reply:
I remember this was a workaround that we introduced at about the time we implemented the named clients. The problem is that `Get(name)` never returns null. If there is no object of the given name, it returns a default instance.

Reply:
I changed this behavior. The option is bounded to its name and the name is compared with the name property. It does not depend on some option property value.

Reply:
I hope the `internal` property won't cause any problems while (de)serializing the `DeliveryOptions` in ASP.NET apps or in the code generator.

Reply:
I will do some testing with the real app.